062-topic-replication-factor-change.md

```diff
@@ -0,0 +1,254 @@
+To further improve fault tolerance, a rack-aware Kafka cluster can be used to distribute topic replicas evenly across data centers in the same region.
+
+Once a topic is created, it is possible to change its replication factor using the `kafka-reassign-partitions.sh` command line tool.
+This operation can be quite complicated on busy clusters with thousands of partitions, as the user have to decide leader or partition movements, without causing uneven load.
+
+The goal of this proposal is to allow Strimzi users to change the topic replication factor by simply updating the `KafkaTopic` resource.
+Unless specified, we always refers to the Unidirectional Topic Operator in this proposal.
+
+## Current situation
+
+When the user updates the `.spec.replicas` field in a `KafkaTopic` resource, the Topic Operator fails the reconciliation, and the resource becomes not ready. 
+A similar error message is reported in Topic Operator logs and `KafkaTopic` resource status:
+
+> Replication factor change not supported, but required for partitions\
+> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,\
+> 22, 23, 24]
+
+The only [documented solution](https://strimzi.io/docs/operators/latest/deploying#proc-changing-topic-replicas-str) is to use the `kafka-reassign-partitions.sh` by spinning up a separate interactive pod.
+This is is far from being user friendly, and the user is in charge of doing the calculations to determine the best broker to host a new replica, or which replica needs to be dropped.
+
+When data loss is not a problem, the user can simply delete and recreate the `KafkaTopic` resource with the desired replication factor.
+This is not really an option for most use cases.
+
+## Motivation
+
+The topic replication factor change feature will provide a much better cloud-native experience for Strimzi users.
+They will be able to change the topic replication factor by simply updating the `KafkaTopic` resource, which may also happen through a GitOps pipeline.
+
+This is a non exhaustive list of use cases that will become significantly easier:
+
+- Set a lower replication factor for non-critical topics to reduce the infrastructure costs (compute, storage and network).
+- Set a lower replication factor to deal with resource shortage (trade-off between fault-tolerance and resource utilization).
+- Set a higher replication factor for critical topics to improve durability and fault tolerance.
+
+## Proposal
+
+The replication factor change feature can be implemented by integrating the Topic Operator with Cruise Control.
+The Cluster Operator is already integrated with Cruise Control, so we can reuse some of the existing logic and classes.
+
+At runtime, the Topic Operator watches and periodically reconcile the configuration of all managed and non-paused `KafkaTopic` resources.
+Among the other configurations, the reconciliation logic also search for `.spec.replicas` update by comparing with the actual topic replication factor.
```

Comment:
`reconciliation logic also search for` ... it's not clear to me what you mean here by "search for".
Each time the user updates the `.spec.replicas` field, an event is raised and the TO watcher is notified I guess.
How does it react to it? Just add to the "queue" a request of changing replication factor? How the user knows which "update" is ongoing?

Reply:
> How does it react to it?

That's the usual reconciliation flow where, after the KT event is notified to the TO, a diff is computed against the actual configuration in Kafka. When the RF change is detected for the first time, then it's put in the `pendingChanges` queue.

> How the user knows which "update" is ongoing?

Looking at the KT status. There is a warning for each phase of the RF change.